Support caching PathImpl instance in the RemoteResourceCache.

mattwoodrow opened the pull request:
#### 7bac2024cb8c59ad65d1a66d849efc90a6cb45af
<pre>
Support caching PathImpl instance in the RemoteResourceCache.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302209">https://bugs.webkit.org/show_bug.cgi?id=302209</a>
&lt;<a href="https://rdar.apple.com/problem/164338336">rdar://problem/164338336</a>&gt;

Reviewed by Kimmo Kinnunen and Simon Fraser.

Serializing paths can be expensive. Having a retained IPC object for paths that
get reused frequently would improve performance.

Path itself isn&apos;t refcounted, and is mutable. The contained PathImpl is
copy-on-write immutable (using DataRef), so this is used as the cached object.

Opts in to the cached paths for svg fill and clip paths, as these are likely to
be rendered multiple times.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/Cairo.cmake:
* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::definitelyEqual const):
(WebCore::Path::ensureProtectedPlatformPathImpl):
(WebCore::Path::setNotTransient):
(WebCore::Path::addPath):
(WebCore::Path::applyElements const):
(WebCore::Path::transform):
(WebCore::Path::platformPath const):
(WebCore::Path::segmentsIfExists const):
(WebCore::Path::contains const):
(WebCore::Path::strokeContains const):
(WebCore::Path::strokeBoundingRect const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/PathImpl.cpp:
(WebCore::lockedPathImplObserversMapSingleton):
(WebCore::PathImpl::~PathImpl):
(WebCore::PathImpl::segments const):
(WebCore::PathImpl::addObserver const):
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::segmentsIfExists const):
(WebCore::PathImpl::isTransient const):
(WebCore::PathImpl::setNotTransient):
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/PlatformPathImpl.h:
* Source/WebCore/platform/graphics/RenderingResource.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
* Source/WebCore/platform/graphics/skia/PathSkia.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::pathOnlyClipping):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::ensurePath):
* Source/WebCore/svg/SVGPathSegList.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.cpp:
(WebKit::RemoteGraphicsContext::clipCachedPath):
(WebKit::RemoteGraphicsContext::fillCachedPath):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cachePathImpl):
(WebKit::RemoteRenderingBackend::releasePathImpl):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cachePathImpl):
(WebKit::RemoteResourceCache::releasePathImpl):
(WebKit::RemoteResourceCache::cachedPathImpl const):
(WebKit::RemoteResourceCache::releaseMemory):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Scripts/webkit/messages.py:
(atomic_object_identifier):
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
(WebKit::RemoteGraphicsContextProxy::clipPath):
(WebKit::RemoteGraphicsContextProxy::fillPath):
(WebKit::RemoteGraphicsContextProxy::recordResourceUse):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemotePathImplIdentifier.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cachePathImpl):
(WebKit::RemoteRenderingBackendProxy::releasePathImpl):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordPathImplUse):
(WebKit::RemoteResourceCacheProxy::willDestroyPathImpl):
(WebKit::RemoteResourceCacheProxy::releaseMemory):
(WebKit::RemoteResourceCacheProxy::disconnect):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Canonical link: <a href="https://commits.webkit.org/306260@main">https://commits.webkit.org/306260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c5782f59e157bb1863d4fa2a395ca4bdc169ba5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/140795 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/13178 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/2354 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/149196 "") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/93883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/142668 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/13889 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/13331 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/108002 "1 flakes") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/93883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/143746 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/10727 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/126039 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/88905 "Passed tests") | | ⏳ 🛠 vision-apple 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/140140 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/10336 "Passed tests") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/9228 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/119582 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/2043 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/151758 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/12865 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/2231 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/116270 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/12880 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/13331 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/116607 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29653 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/12607 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/122681 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/68005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/12910 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/2147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/12647 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/76608 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/12846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/12691 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/Path.h:
```diff
@@ -114,6 +114,10 @@ class Path {
+    WEBCORE_EXPORT RefPtr<const PathImpl> asImpl() const;
+
+    bool isTransient() const { return m_isTransient; }
+    void setCacheable() { m_isTransient = false; }
```
I can see why you did it this way, but it might be clearer if these two functions used the same name.

smfr commented on Source/WebCore/platform/graphics/PathImpl.h:
```diff
@@ -86,8 +90,15 @@ class PathImpl : public ThreadSafeRefCounted<PathImpl> {
 protected:
     PathImpl() = default;
+
+    mutable WeakHashSet<RenderingResourceObserver> m_observers;
```
Since cached paths are rare, bloating PathImpl with this hash set may not be the best idea. How about a side table, only accessed if the path is cacheable?

smfr commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.h:
```diff
@@ -95,6 +96,7 @@ class RemoteGraphicsContext : public IPC::StreamMessageReceiver, public CanMakeW
     void clipToImageBuffer(WebCore::RenderingResourceIdentifier, const WebCore::FloatRect& destinationRect);
     void clipOutToPath(const WebCore::Path&);
     void clipPath(const WebCore::Path&, WebCore::WindRule);
+    void clipCachedPath(RemotePathImplIdentifier, WebCore::WindRule);
```
Why does this take not a `WebCore::RenderingResourceIdentifier` like the others?

mattwoodrow replied:
@kkinnunen-apple has been actively converting the resources to use their own (WebKit, not WebCore) identifiers.

kkinnunen-apple commented on Source/WebCore/platform/graphics/Path.h:
```diff
@@ -114,6 +114,10 @@ class Path {
     WEBCORE_EXPORT void ensureImplForTesting();
+    WEBCORE_EXPORT RefPtr<const PathImpl> asImpl() const;
+
+    bool isTransient() const { return m_isTransient; }
```
Immutable copy-on-write path cannot really switch transparently between transient and not-transient.
Effectively the Path becomes refptr of impl plus a vehicle to communicate draw options to the draws.
It's kind of detrimental of the possible future case where we'd change Path to be more similar to other resources, e.g. immutable direct-refcounted object. Granted, this is speculation, but adding fields in the wrapper holder is taking an opposing direction.

kkinnunen-apple commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
```diff
@@ -158,11 +160,15 @@ auto LegacyRenderSVGResourceClipper::pathOnlyClipping(GraphicsContext& context,
             result.add(ApplyResult::ClipContainsRendererContent);
     }
 
+    clipPath.setNotTransient();
```
I have quite strong feelings about the caching of the rendering resources and identifiers and such.

In general, I think we should move into direction where the caching of gradients, filters, etc. should happen at the display list scale.

For your use-case, could we cache somehow a maximal display list instead?

So for here, for example, it's not evident that strictly bit-equal path is created ever again. It's an object that gets created during each pass. Thus it's not very nice, as you need to start hashing the contents..

kkinnunen-apple reviewed (changes requested):
Marking this as request changes just to discuss this, still.
There's major amount of problems with current resource handling, identifiers, etc. Thread safety and such, which we have already discussed.  Maybe there could be a more scalable approach..

kkinnunen-apple commented on Source/WebCore/platform/graphics/PathImpl.cpp:
```diff
@@ -31,6 +31,38 @@ namespace WebCore {
+static PathImplObserversMap* gPathImplObserversMap;
+static PathImplObserversMap& pathImplObserversMapSingleton()
+{
+    if (!gPathImplObserversMap)
```
this cannot be non-thread safe..

kkinnunen-apple commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
```diff
@@ -136,19 +136,21 @@ auto LegacyRenderSVGResourceClipper::pathOnlyClipping(GraphicsContext& context,
 
     // Transform path by animatedLocalTransform.
-    clipPath.transform(animatedLocalTransform);
+    AffineTransform oldCTM = context.getCTM();
```
Note: getting CTM for restore purposes kind of breaks display list recording, so the context-agnostic code becomes bound to non-display list drawing code only.
As opposed to save+restore.
I think we should have optimised-enough code to support save+restore to work only on the changed properties. If not, we should.

mattwoodrow replied:
I don't think we do have that, unfortunately. We just forward the save/restore calls as-is across IPC.

mattwoodrow replied:
Probably not a big deal to do that here though, was just trying to do the most efficient thing.

mattwoodrow replied:
Whoops, nope. Adding save/restore removes the clip as well as the transform. We want just the clip to be transformed, but then we want untransformed (but clipped!) for following commands.

kkinnunen-apple commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
```diff
@@ -158,11 +160,15 @@ auto LegacyRenderSVGResourceClipper::pathOnlyClipping(GraphicsContext& context,
 
+    clipPath.setNotTransient();
     context.clipPath(clipPath, clipRule);
+    context.setCTM(oldCTM);
```
From API perspective I'd much rather see something like:

* Work towards removing the "is transient" resource state on the grounds that typically at the general purpose code creation site, it's not known if the resource is actually long-lived or not. Current gradient is transient state is only to replicate the original state without binding the state to presence of the identifier. It's not a real design in that way
* Work towards holding display lists
* Work towards adding per draw operation transiency state. Here, it would be `context.clipPath(clipPath, clipRule, ClipPathOptions { PersistentPath }`.
* If nothing else works, add "is transient" state to the resources


What I like about this change is that the path is not transformed, rather the context transform.

What I don't like about this change is that the SVG patterns and such are the *prime* use-case to record the whole subshapes into the display lists. I would imagine this was the projected use-case for the tags. If we work on adding caching to individual resources, we leave much on the table.

mattwoodrow replied:
I think in the case of Paths, the creation site does know (think?) it'll be retained, and it's harder to tell at the call site.
I think this caller is the incorrect one, as we don't really know if we've take the PathImpl directly from the SVG itself, or if we've since mutated the path (and triggered a clone of the PathImpl).

I think moving this `setNotTransient` call to be on the inner SVG data (like the other caller of this function) would be more consistent.

I also think the transient flag should probably be on the PathImpl, such that DataRef clones for mutation reset the value.

kkinnunen-apple commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
```diff
@@ -136,19 +136,21 @@ auto LegacyRenderSVGResourceClipper::pathOnlyClipping(GraphicsContext& context,
     }
```
With all below and above said, with the code here I can see how "just record a display list and invalidate it when something changes" is easier said than done.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
```diff
@@ -56,6 +57,10 @@ class RemoteResourceCache {
 
+    bool cachePathImpl(RemotePathImplIdentifier, Ref<WebCore::PathImpl>&&);
+    bool releasePathImpl(RemotePathImplIdentifier);
+    RefPtr<WebCore::PathImpl> cachedPathImpl(RemotePathImplIdentifier) const;
```
While caching pathimpls, I think it's still consistent to return Paths. This way you can set up the persistency state correctly for the Path.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.cpp:
```diff
@@ -341,6 +341,15 @@ void RemoteGraphicsContext::clipPath(const Path& path, WindRule rule)
+    MESSAGE_CHECK(pathImpl);
+
+    Path path(pathImpl.releaseNonNull());
+    clipPath(path, rule);
```
It has bitten me before calling a public interface function as part of the implementation.
* If somebody adds a preamble to every public interface function, then you end up running then preamble twice.
* Iit kind of gives an impression that all public functions validate their parameters, but then it kind of might have a temptation to not validate the overlapping parameters

consider dispatching directly to the implementation, e.g. `context().clipPath(path, rule);`.

kkinnunen-apple commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -142,6 +142,12 @@ RefPtr<const PathImpl> Path::asImpl() const
 
+void Path::setNotTransient() const
+{
+    if (auto impl = asImpl())
```
It's bit of a disappointment wrt immutability, consistency, equality, the way paths move in and out of implementation state.

can you make it ensureImpl?
Is it possible to make it non-const or copy constructor (Path::createTransient(otherpath)? 

```
Path a;
a.setNotTransient();
for (int i = 0; i<10; ++i)
    a.addLineTo(i,i);
Path b;
for (int i = 0; i<10; ++i)
    b.addLineTo(i,i);
b.setNotTransient();

//is a same as b?
```

kkinnunen-apple reviewed (approved):
I'm a bit disappointed that it's complex work that doesn't necessarily take it into more consistent direction. I don't have cycles for anything better..

mattwoodrow committed:
Support caching PathImpl instance in the RemoteResourceCache.
https://bugs.webkit.org/show_bug.cgi?id=302209
<rdar://problem/164338336>

Reviewed by Kimmo Kinnunen and Simon Fraser.

Serializing paths can be expensive. Having a retained IPC object for paths that
get reused frequently would improve performance.

Path itself isn't refcounted, and is mutable. The contained PathImpl is
copy-on-write immutable (using DataRef), so this is used as the cached object.

Opts in to the cached paths for svg fill and clip paths, as these are likely to
be rendered multiple times.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/Cairo.cmake:
* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::definitelyEqual const):
(WebCore::Path::ensureProtectedPlatformPathImpl):
(WebCore::Path::setNotTransient):
(WebCore::Path::addPath):
(WebCore::Path::applyElements const):
(WebCore::Path::transform):
(WebCore::Path::platformPath const):
(WebCore::Path::segmentsIfExists const):
(WebCore::Path::contains const):
(WebCore::Path::strokeContains const):
(WebCore::Path::strokeBoundingRect const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/PathImpl.cpp:
(WebCore::lockedPathImplObserversMapSingleton):
(WebCore::PathImpl::~PathImpl):
(WebCore::PathImpl::segments const):
(WebCore::PathImpl::addObserver const):
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::segmentsIfExists const):
(WebCore::PathImpl::isTransient const):
(WebCore::PathImpl::setNotTransient):
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/PlatformPathImpl.h:
* Source/WebCore/platform/graphics/RenderingResource.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
* Source/WebCore/platform/graphics/skia/PathSkia.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::pathOnlyClipping):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::ensurePath):
* Source/WebCore/svg/SVGPathSegList.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.cpp:
(WebKit::RemoteGraphicsContext::clipCachedPath):
(WebKit::RemoteGraphicsContext::fillCachedPath):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cachePathImpl):
(WebKit::RemoteRenderingBackend::releasePathImpl):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cachePathImpl):
(WebKit::RemoteResourceCache::releasePathImpl):
(WebKit::RemoteResourceCache::cachedPathImpl const):
(WebKit::RemoteResourceCache::releaseMemory):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Scripts/webkit/messages.py:
(atomic_object_identifier):
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
(WebKit::RemoteGraphicsContextProxy::clipPath):
(WebKit::RemoteGraphicsContextProxy::fillPath):
(WebKit::RemoteGraphicsContextProxy::recordResourceUse):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemotePathImplIdentifier.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cachePathImpl):
(WebKit::RemoteRenderingBackendProxy::releasePathImpl):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordPathImplUse):
(WebKit::RemoteResourceCacheProxy::willDestroyPathImpl):
(WebKit::RemoteResourceCacheProxy::releaseMemory):
(WebKit::RemoteResourceCacheProxy::disconnect):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Canonical link: https://commits.webkit.org/306260@main
